feat: Handle CouchDB HTTP 403 on all routes

H--o-l opened the pull request:
Since CouchDB v3.4.0, there has been a new "Lockout" feature, i.e., a rate limit
on tuples (IP, login) after multiple authentication failures. It's highlighted
in the release note:
https://docs.couchdb.org/en/stable/whatsnew/3.4.html#version-3-4-0
(see the second-to-last bullet point of the Highlights section).

As the following upstream discussion shows, this CouchDB feature adds a new case
of HTTP 403 possible on all routes:
https://github.com/apache/couchdb/issues/5315#issuecomment-2427009998

This commit catches the 403 on all routes. As some routes were already catching
403 for other reasons, the exception message on these routes is changed from
their previous message to `"Access forbidden: {reason}"` where `reason` is
either the `reason` returned by CouchDB in the JSON body of the answer, or if
it doesn't exist, by the `message` of aiohttp ClientResponseError.

Note that for the streamed version of the route `/_changes`, I wasn't able
to fill any reason, so the exception message will be `"Access forbidden: None"`.

I manually tested with `await couchdb.info()`, it returns the following:

```
> await couchdb.info()
...
aiocouch.exception.UnauthorizedError: Invalid credentials
> await couchdb.info()  # <=== Lockout
...
aiocouch.exception.ForbiddenError: Access forbidden: Account is temporarily
locked due to multiple authentication failures
```

Closes #55

H--o-l committed:
feat: Handle CouchDB HTTP 403 on all routes

Since CouchDB v3.4.0, there has been a new "Lockout" feature, i.e., a rate limit
on tuples (IP, login) after multiple authentication failures. It's highlighted
in the release note:
https://docs.couchdb.org/en/stable/whatsnew/3.4.html#version-3-4-0
(see the second-to-last bullet point of the Highlights section).

As the following upstream discussion shows, this CouchDB feature adds a new case
of HTTP 403 possible on all routes:
https://github.com/apache/couchdb/issues/5315#issuecomment-2427009998

This commit catches the 403 on all routes. As some routes were already catching
403 for other reasons, the exception message on these routes is changed from
their previous message to `"Access forbidden: {reason}"` where `reason` is
either the `reason` returned by CouchDB in the JSON body of the answer, or if
it doesn't exist, by the `message` of aiohttp ClientResponseError.

Note that for the streamed version of the route `/_changes`, I wasn't able
to fill any reason, so the exception message will be `"Access forbidden: None"`.

I manually tested with `await couchdb.info()`, it returns the following:

```
> await couchdb.info()
...
aiocouch.exception.UnauthorizedError: Invalid credentials
> await couchdb.info()  # <=== Lockout
...
aiocouch.exception.ForbiddenError: Access forbidden: Account is temporarily
locked due to multiple authentication failures
```

Closes #55

H--o-l commented on aiocouch/remote.py:
```diff
@@ -179,6 +186,12 @@ async def _streamed_request(
+            if not resp.ok:
+                # Save the reason for later if any
+                try:
+                    self.failure_reason = (await resp.json())["reason"]
```
I'm unfamiliar enough with aiohttp streamed request to know if I'm doing something wrong here. 

Please give your opinion on it I would be grateful.

H--o-l commented on aiocouch/remote.py:
```diff
@@ -203,12 +217,14 @@ async def close(self) -> None:
         await asyncio.sleep(0.250 if has_ssl_conn else 0)
 
     @raises(401, "Invalid credentials")
+    @raises(403, "Access forbidden: {reason}")
```
I only manually tested this one with a real CouchDB, not the other routes.

bmario commented on aiocouch/remote.py:
```diff
@@ -160,6 +161,12 @@ async def _request(
         async with self._http_session.request(
             method, url=f"{self._server}{path}", **kwargs
         ) as resp:
+            if not resp.ok:
+                # Save the reason for later if any
+                try:
+                    self.failure_reason = (await resp.json())["reason"]
+                except:
+                    self.failure_reason = None
             resp.raise_for_status()
```
Okay, my bad. I only quickly checked the documentation and assumed that the `message` field in the ClientResponseError is the body of the response. So I assumed, that there is no need to pass the response body around.

However, I don't like the indirect setting of failure_reason through the endpoint instances. Instead, I'd extend the ClientResponseError and add a new field reason, which would be set from the response JSON in its constructor. And then throw this exception manually instead of the raise_for_status call.

H--o-l replied:
Noted, I'm on it but I didn't had the time to finish today, I will finish next week. Have a good weekend!

H--o-l replied:
> So I assumed, that there is no need to pass the response body around.

I checked the [aiohttp v3.9.5 code](https://github.com/aio-libs/aiohttp/blob/b844d4293a5bd4921bd7267550294122e83617a8/aiohttp/client_reqrep.py#L1065-L1076), and I didn't find a way to add the body reason to the exception.

> However, I don't like the indirect setting of failure_reason through the endpoint instances. Instead, I'd extend the ClientResponseError and add a new field reason, which would be set from the response JSON in its constructor. And then throw this exception manually instead of the raise_for_status call.

I updated the PR with your suggestion, thanks.

When raising the exception manually instead of calling aiohttp `raise_for_status()`, I copied the whole code of `raise_for_status()` inside aiocouch. I let you see what it gives inside the PR diff but, to be noted that I don't like it, as this upstream code could change without aiocouch being updated. For example, between aiohttp v3.9.5 and aiohttp v3.10.10 that code already changed.
In v3.9.5:
```py
    def raise_for_status(self) -> None:
        if not self.ok:
            # reason should always be not None for a started response
            assert self.reason is not None
            self.release()
            raise ClientResponseError(
                self.request_info,
                self.history,
                status=self.status,
                message=self.reason,
                headers=self.headers,
            )
```
In v3.10.10:
```py
   def raise_for_status(self) -> None:
        if not self.ok:
            # reason should always be not None for a started response
            assert self.reason is not None

            # If we're in a context we can rely on __aexit__() to release as the
            # exception propagates.
            if not self._in_context:
                self.release()

            raise ClientResponseError(
                self.request_info,
                self.history,
                status=self.status,
                message=self.reason,
                headers=self.headers,
            )
```

Solutions I see:

1. Don't copy-paste this code and just manually raise `ClientResponseError()`; that could let some aiohttp "context" in the wild, ending in a memory leak (I guess).
2. don't get the `reason` from the CouchDB body answer and just use the aiohttp.ClientResponseError exception message, but it's really vague.
3. contribute to aiohttp so the body of the request is set inside `aionttp.ClientResponseError()`, but I don't have the shoulders or the time to do that.

If you agree with me that the current code is not maintainable, I think solution `2.` is the best in the end. What do you think?

bmario commented:
Can you please rebase onto master, I updated the CI there.

H--o-l commented:
> Can you please rebase onto master, I updated the CI there.

Hi @bmario, it's done!

bmario commented:
Thanks, there seems to be an issue with sorting imports. Can you iron these out with `isort`?

H--o-l commented:
> Thanks, there seems to be an issue with sorting imports. Can you iron these out with isort?

Thanks for the lead, done.

To be noted that the current version of this PR doesn't seem very maintainable for me when upgrading aiohttp. I discussed it here:
 https://github.com/metricq/aiocouch/pull/56#discussion_r1821237749

bmario commented:
Well. I don't think it would actually a problem, but  we could just call raise_from_status and wrap that call in a try-except. Handling the reason extraction in except path and raise a new exception with that attached.

H--o-l commented:
Hey @bmario!

> Well. I don't think it would actually a problem, but we could just call raise_from_status and wrap that call in a try-except. Handling the reason extraction in except path and raise a new exception with that attached.

Great idea, thanks.

I pushed it, slightly updated the commit message, fixed the linting issue, and removed the `reason = (await resp.json())["reason"]` part in the cases of a "streamed request" (i.e. `GET _changes` with `feed = continuous`). Indeed, trying to get the `reason` in that special case was freezing the connection. Instead, I set the reason to `None,` which I think is OK in that case, but don't hesitate to challenge me.

Note that codecov shows a small drop in code coverage, but I don't know how to fix it: I'm not used to code coverage.
If you think it's worth spending time on it, tell me and I will try.

H--o-l commented:
Hi @bmario! 

What do you think of this change? 
I think the PR is ready to be merged, if you have a little time to check it :crossed_fingers:

H--o-l commented:
Hey @bmario!

Small ping: merging this PR would clean my code handling that case, and maybe clarify this error case for other users too, if you have time to re-check it, it would be cool :pray:

bmario reviewed (changes requested):
In general, this look good to me, but the name of the exception.
I think it is confusing to have the same name as the aiohttp exception. This makes it hard to see at a glance, what exception this is. Also "Client" corresponds to the aiohttp.ClientSession. So in the context of aiocouch, that makes less sense. 

I'd suggest `RemoteResponseError`.

H--o-l commented:
@bmario thanks!
